crates/bevy_ui/src/render/mod.rs

```diff
@@ -173,29 +184,81 @@ pub fn extract_uinodes(
             Option<&UiImage>,
             &ComputedVisibility,
             Option<&CalculatedClip>,
+            Option<&Handle<TextureAtlas>>,
+            Option<&UiTextureAtlasSprite>,
         )>,
     >,
 ) {
     extracted_uinodes.uinodes.clear();
     for (stack_index, entity) in ui_stack.uinodes.iter().enumerate() {
-        if let Ok((uinode, transform, color, maybe_image, visibility, clip)) =
-            uinode_query.get(*entity)
+        if let Ok((
+            uinode,
+            transform,
+            color,
+            maybe_image,
+            visibility,
+            clip,
+            maybe_texture_atlas,
+            maybe_atlas_sprite,
+        )) = uinode_query.get(*entity)
         {
             // Skip invisible and completely transparent nodes
             if !visibility.is_visible() || color.0.a() == 0.0 {
                 continue;
             }
 
-            let (image, flip_x, flip_y) = if let Some(image) = maybe_image {
-                // Skip loading images
-                if !images.contains(&image.texture) {
+            let atlas_details = if let Some(atlas_sprite) = maybe_atlas_sprite {
+                if let Some(texture_atlas_handle) = maybe_texture_atlas {
+                    if let Some(texture_atlas) = texture_atlases.get(texture_atlas_handle) {
+                        let atlas_rect = *texture_atlas
+                                .textures
+                                .get(atlas_sprite.0.index)
+                                .unwrap_or_else(|| {
+                                    panic!(
+                                        "Sprite index {:?} does not exist for texture atlas handle {:?}.",
+                                        atlas_sprite.0.index,
+                                        texture_atlas_handle.id(),
+                                    )
+                                });
+                        Some(ExtractAtlasDetails {
+                            rect: atlas_rect,
+                            size: texture_atlas.size,
+                            image: texture_atlas.texture.clone(),
+                            flip_x: atlas_sprite.0.flip_x,
+                            flip_y: atlas_sprite.0.flip_y,
+                        })
+                    } else {
+                        // Atlas not present in assets resource (should this warn the user?)
+                        None
+                    }
+                } else {
+                    warn!("UI Texture Atlas is missing a Handle<TextureAtlas> Component");
                     continue;
                 }
-                (image.texture.clone_weak(), image.flip_x, image.flip_y)
             } else {
-                (DEFAULT_IMAGE_HANDLE.typed().clone_weak(), false, false)
+                // Not using texture atlas for this node
+                None
             };
 
```

Comment:
Instead of adding a `uv_rect` field, set the values for `atlas_size` and `rect` using:
```
let mut rect = texture_atlas.textures[atlas_image.index];
let scale = uinode.size() / rect.size();
rect.min *= scale;
rect.max *= scale;
let atlas_size = texture_atlas.size * scale;
```

Reply:
I saw text does this but I couldn't understand the math there.  They used the inverse scale and also modified the transform by the same value.  When I tried the same thing here I couldn't get it to work but I can try again.

Reply:
Update: You are 100% correct! Changed!